File: struct-test/structTest.go
package main

import "fmt"

// 定義一個結構體
type Point struct {
    X int
    Y int
}

func main() {
    fmt.Println("=== Go 值類型結構體測試 ===")
    
    // 測試值類型賦值
    point1 := Point{X: 1, Y: 2}
    point2 := point1 // 創建副本，而不是引用
    
    // 修改 point2
    point2.X = 10
    
    fmt.Println("\n修改 point2 後：")
    fmt.Printf("point1: %+v\n", point1) // 不會被修改
    fmt.Printf("point2: %+v\n", point2)
    
    // 測試指針引用
    fmt.Println("\n=== Go 指針引用測試 ===")
    
    point3 := &Point{X: 1, Y: 2}
    point4 := point3 // 共享同一個指針
    
    point4.X = 10
    
    fmt.Println("\n修改 point4 後：")
    fmt.Printf("point3: %+v\n", *point3) // 會被修改
    fmt.Printf("point4: %+v\n", *point4)
    
    // 測試數組中的結構體
    fmt.Println("\n=== Go 數組中的結構體測試 ===")
    
    points := []Point{
        {X: 1, Y: 1},
        {X: 2, Y: 2},
        {X: 3, Y: 3},
    }
    
    firstPoint := points[0] // 創建副本
    firstPoint.X = 10
    
    fmt.Println("\n修改 firstPoint 後：")
    fmt.Printf("數組中的第一個點：%+v\n", points[0]) // 不會被修改
    fmt.Printf("firstPoint：%+v\n", firstPoint)
} 

File: struct-test/structTest.js
// JavaScript 值類型結構體測試

// 定義一個"結構體"（實際上是對象）
class Point {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}

// 測試函數：驗證對象引用特性
function testObjectReference() {
    console.log('=== JavaScript 對象引用測試 ===');
    
    // 創建一個"結構體"實例
    const point1 = new Point(1, 2);
    
    // 賦值給另一個變量（創建引用）
    const point2 = point1;
    
    // 修改 point2
    point2.x = 10;
    
    // 檢查 point1 是否也被修改
    console.log('修改 point2 後：');
    console.log('point1:', point1); // 會被修改，因為是引用
    console.log('point2:', point2);
    console.log('point1 === point2:', point1 === point2); // true，同一個對象
}

// 測試函數：嘗試模擬值類型行為
function testValueTypeCopy() {
    console.log('\n=== JavaScript 值類型複製模擬測試 ===');
    
    // 創建一個"結構體"實例
    const point1 = new Point(1, 2);
    
    // 通過複製屬性創建新對象
    const point2 = new Point(point1.x, point1.y);
    
    // 修改 point2
    point2.x = 10;
    
    // 檢查 point1 是否被修改
    console.log('修改 point2 後：');
    console.log('point1:', point1); // 不會被修改，因為是獨立對象
    console.log('point2:', point2);
    console.log('point1 === point2:', point1 === point2); // false，不同對象
}

// 測試函數：數組中的對象引用
function testArrayOfObjects() {
    console.log('\n=== JavaScript 數組中的對象引用測試 ===');
    
    // 創建包含多個點的數組
    const points = [
        new Point(1, 1),
        new Point(2, 2),
        new Point(3, 3)
    ];
    
    // 獲取第一個點的引用
    const firstPoint = points[0];
    
    // 修改引用的點
    firstPoint.x = 10;
    
    // 檢查數組中的原始對象是否被修改
    console.log('修改 firstPoint 後：');
    console.log('數組中的第一個點：', points[0]); // 會被修改
    console.log('firstPoint：', firstPoint);
}

// 運行所有測試
testObjectReference();
testValueTypeCopy();
testArrayOfObjects(); 